ports/raspberrypi/common-hal/rp2pio/StateMachine.c

```diff
@@ -343,15 +346,34 @@ bool rp2pio_statemachine_construct(rp2pio_statemachine_obj_t *self,
 
     sm_config_set_wrap(&c, wrap_target, wrap);
     sm_config_set_in_shift(&c, in_shift_right, auto_push, push_threshold);
-    sm_config_set_out_shift(&c, out_shift_right, auto_pull, pull_threshold);
+    #if PICO_PIO_VERSION > 0
+    sm_config_set_in_pin_count(&c, in_pin_count);
+    #endif
 
-    enum pio_fifo_join join = PIO_FIFO_JOIN_NONE;
-    if (!rx_fifo) {
-        join = PIO_FIFO_JOIN_TX;
-    } else if (!tx_fifo) {
-        join = PIO_FIFO_JOIN_RX;
+    sm_config_set_out_shift(&c, out_shift_right, auto_pull, pull_threshold);
+    sm_config_set_out_pin_count(&c, out_pin_count);
+
+    sm_config_set_set_pin_count(&c, set_pin_count);
+
+    enum pio_fifo_join join =
+        fifo_type != PIO_FIFO_JOIN_AUTO ? fifo_type
+        : !rx_fifo ? PIO_FIFO_JOIN_TX
+        : !tx_fifo ? PIO_FIFO_JOIN_RX
+        : PIO_FIFO_JOIN_NONE;
+    self->fifo_depth = (join == PIO_FIFO_JOIN_TX || join == PIO_FIFO_JOIN_RX) ? 8
+        #if PICO_PIO_VERSION > 0
+        : (join == PIO_FIFO_JOIN_PUTGET) ? 0
+        #endif
+        : 4;
```

Comment:
Please add parens or switch to if statements. Nested ternary is very hard for me to read.